lib/assertions.js

```diff
@@ -344,9 +344,89 @@ module.exports = expect => {
           this.outdentLines();
         });
       }
-      propertyNames.forEach(propertyName => {
-        expect(subject, '[not] to have [own] property', String(propertyName));
-      });
+
+      if (expect.flags.only) {
+        if (expect.flags.not) {
+          expect.errorMode = 'bubble';
+          expect.fail(
+            'The "not" flag cannot be used together with "to only have properties".'
+          );
+        }
+        if (expect.flags.own) {
+          expect.errorMode = 'bubble';
+          expect.fail(
+            'The "own" flag cannot be used together with "to only have properties".'
+          );
+        }
+        const subjectType = expect.subjectType;
+        const subjectKeys = subjectType.getKeys(subject).filter(
+          key =>
+            utils.numericalRegExp.test(key) ||
+            typeof key === 'symbol' ||
+            // include only those keys whose value is not undefined
+            typeof subjectType.valueForKey(subject, key) !== 'undefined'
+        );
+
+        expect.withError(
+          () => {
+            expect(
+              subjectKeys.length === propertyNames.length,
+              '[not] to be truthy'
+            );
+          },
+          () => {
+            expect.fail({
+              diff:
+                !expect.flags.not &&
+                ((output, diff, inspect, equal) => {
+                  output.inline = true;
+
+                  const keyInValue = {};
+                  propertyNames.forEach(key => {
+                    keyInValue[key] = true;
+                  });
+
+                  subjectType.prefix(output, subject);
+                  output.nl().indentLines();
+
+                  subjectKeys.forEach((key, index) => {
+                    const propertyOutput = subjectType.property(
+                      output.clone(),
+                      key,
+                      inspect(subjectType.valueForKey(subject, key))
+                    );
+                    const delimiterOutput = subjectType.delimiter(
+                      output.clone(),
+                      index,
+                      subjectKeys.length
+                    );
+
+                    output
+                      .i()
+                      .block(function() {
+                        this.append(propertyOutput).amend(delimiterOutput);
+                        if (!keyInValue[key]) {
+                          this.sp().annotationBlock(function() {
+                            this.error('should be removed');
+                          });
+                        }
+                      })
+                      .nl();
+                  });
```

Comment:
You should iterate over the `propertyNames` that are missing in `subject` here and show that in the diff as well.

Reply:
Great shout! Addressed in the latest push :)